File: frontend/src/api/customer.ts
import api from './index'

export interface Customer {
  id: number
  name: string
  code?: string
  industry?: string
  sales_rep_id: number
  tier_level: string
  annual_consumption: number
  status: string
  contact_person?: string
  contact_phone?: string
  contact_email?: string
  address?: string
  remark?: string
  created_at: string
  updated_at: string
}

export interface CustomerQuery {
  page?: number
  size?: number
  keyword?: string
  sales_rep_ids?: number[]
  industries?: string[]
  status?: string[]
  tier_levels?: string[]
  annual_consumption_min?: number
  annual_consumption_max?: number
  created_at_start?: string
  created_at_end?: string
  updated_at_start?: string
  updated_at_end?: string
  sort_field?: string
  sort_desc?: boolean
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  size: number
  pages: number
}

export const customerApi = {
  async list(query: CustomerQuery): Promise<PaginatedResponse<Customer>> {
    const response = await api.get<PaginatedResponse<Customer>>('/customers', {
      params: query
    })
    return response.data
  },

  async get(id: number): Promise<Customer> {
    const response = await api.get<Customer>(`/customers/${id}`)
    return response.data
  },

  async create(data: Partial<Customer>): Promise<Customer> {
    const response = await api.post<Customer>('/customers', data)
    return response.data
  },

  async update(id: number, data: Partial<Customer>): Promise<Customer> {
    const response = await api.put<Customer>(`/customers/${id}`, data)
    return response.data
  },

  async delete(id: number): Promise<void> {
    await api.delete(`/customers/${id}`)
  }
}

File: backend/app/schemas/customer.py
from pydantic import BaseModel, ConfigDict, EmailStr, Field
from typing import Optional, List
from datetime import datetime


class CustomerCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    code: Optional[str] = Field(None, max_length=50)
    industry: Optional[str] = Field(None, max_length=100)
    sales_rep_id: int
    tier_level: Optional[str] = Field("D", pattern="^[A-D]$")
    annual_consumption: Optional[float] = Field(0, ge=0)
    status: Optional[str] = Field("active", pattern="^(active|inactive)$")
    contact_person: Optional[str] = Field(None, max_length=100)
    contact_phone: Optional[str] = Field(None, pattern=r"^\d{11}$")
    contact_email: Optional[EmailStr] = None
    address: Optional[str] = Field(None, max_length=500)
    remark: Optional[str] = Field(None, max_length=1000)


class CustomerUpdateRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    code: Optional[str] = Field(None, max_length=50)
    industry: Optional[str] = Field(None, max_length=100)
    tier_level: Optional[str] = Field(None, pattern="^[A-D]$")
    annual_consumption: Optional[float] = Field(None, ge=0)
    status: Optional[str] = Field(None, pattern="^(active|inactive)$")
    contact_person: Optional[str] = Field(None, max_length=100)
    contact_phone: Optional[str] = Field(None, pattern=r"^\d{11}$")
    contact_email: Optional[EmailStr] = None
    address: Optional[str] = Field(None, max_length=500)
    remark: Optional[str] = Field(None, max_length=1000)


class CustomerResponse(BaseModel):
    id: int
    name: str
    code: Optional[str]
    industry: Optional[str]
    sales_rep_id: int
    tier_level: str
    annual_consumption: float
    status: str
    contact_person: Optional[str]
    contact_phone: Optional[str]
    contact_email: Optional[str]
    address: Optional[str]
    remark: Optional[str]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
